src/metabase/lib/schema/aggregation.cljc

```diff
@@ -5,29 +5,71 @@
    [metabase.util.malli.registry :as mr]))
 
 ;; count has an optional expression arg
-(mbql-clause/define-catn-mbql-clause :count
-  [:expression [:? [:ref ::expression/number]]])
+(mbql-clause/define-catn-mbql-clause :count :- :type/Integer
+  [:expression [:? [:schema [:ref ::expression/number]]]])
 
-(defmethod expression/type-of* :count
-  [[_tag _opts expr]]
-  (if-not expr
-    :type/Integer
-    (expression/type-of expr)))
+(mbql-clause/define-tuple-mbql-clause :avg :- :type/Float
+  [:schema [:ref ::expression/number]])
+
+(mbql-clause/define-tuple-mbql-clause :distinct :- :type/Integer
+  [:schema [:ref ::expression/expression]])
+
+(mbql-clause/define-tuple-mbql-clause :count-where :- :type/Integer
+  [:schema [:ref ::expression/boolean]])
+
+(mbql-clause/define-tuple-mbql-clause :max
```

Comment:
Cannot these clauses constrained to return numbers?

Reply:
IIUC That's what `expression/register-type-of-first-arg` will do, if you pass an `Integer` field to `max`, that clause will be constrained to `Integer`.